infrastructure/code/backend_service.go

```diff
@@ -349,3 +351,110 @@ func (s *SnykCodeHTTPClient) checkResponseCode(r *http.Response) error {
 
 	return errors.New("Unexpected response code: " + r.Status)
 }
+
+type AutofixStatus struct {
+	message string
+}
+
+func (s *SnykCodeHTTPClient) RunAutofix(
+	ctx context.Context,
+	options AutofixOptions,
+) ([]snyk.AutofixSuggestion,
+	AutofixStatus,
+	error,
+) {
+	method := "code.RunAutofix"
+	span := s.instrumentor.StartSpan(ctx, method)
+	defer s.instrumentor.Finish(span)
+
+	requestId, err := performance2.GetTraceId(span.Context())
+	if err != nil {
+		log.Err(err).Str("method", method).Msg("Failed to obtain request id. " + err.Error())
+		return nil, AutofixStatus{}, err
+	}
+	log.Debug().Str("method", method).Str("requestId", requestId).Msg("API: Retrieving analysis for bundle")
+	defer log.Debug().Str("method", method).Str("requestId", requestId).Msg("API: Retrieving analysis done")
+
+	requestBody, err := s.autofixRequestBody(&options)
+	if err != nil {
+		log.Err(err).Str("method", method).Str("requestBody", string(requestBody)).Msg("error creating request body")
+		return nil, AutofixStatus{}, err
+	}
+
+	responseBody, err := s.doCall(span.Context(), "POST", "/autofix/suggestions", requestBody)
+	failed := AutofixStatus{message: "FAILED"}
+	if err != nil {
+		log.Err(err).Str("method", method).Str("responseBody", string(responseBody)).Msg("error response from analysis")
+		return nil, failed, err
+	}
+
+	var response AutofixResponse
+	err = json.Unmarshal(responseBody, &response)
+	if err != nil {
+		log.Err(err).Str("method", method).Str("responseBody", string(responseBody)).Msg("error unmarshalling")
+		return nil, failed, err
+	}
+
+	log.Debug().Str("method", method).Str("requestId", requestId).Msgf("Status: %s", response.Status)
+
+	if response.Status == failed.message {
+		log.Err(err).Str("method", method).Str("responseStatus", response.Status).Msg("analysis failed")
+		return nil, failed, SnykAutofixFailedError{Msg: string(responseBody)}
+	}
+
+	if response.Status == "" {
+		log.Err(err).Str("method", method).Str("responseStatus", response.Status).Msg("unknown response status (empty)")
+		return nil, failed, SnykAutofixFailedError{Msg: string(responseBody)}
+	}
+	status := AutofixStatus{message: response.Status}
+	if response.Status != completeStatus {
+		return nil, status, nil
+	}
+
+	suggestions := response.toAutofixSuggestions(options.filePath)
+	return suggestions, AutofixStatus{message: response.Status}, nil
+}
+
+func (s *SnykCodeHTTPClient) autofixRequestBody(options *AutofixOptions) ([]byte, error) {
+	orgName := "unknown"
+	if config.CurrentConfig().GetOrganization() != "" {
+		orgName = config.CurrentConfig().GetOrganization()
+	}
+
+	issueData, ok := options.issue.AdditionalData.(IssueData)
```

Comment:
You should be careful to depend on anything from `snyk.Issue`. Since it's a domain object, any change to it will impact the autofix behaviour. For instance, `AdditionalData` was meant as a view model for UI so far only. If we decide to change our UI, ruleId dependent logic may break here by mistake. 

As a solution I'd suggest to have own mapping between `snyk.Issue` and the Code issue model that stores its own representation that's needed for autofix. E.g. for ruleId, line number, as they come from `options.issue`.

It's more or less ok for now, but you/we should watch out for regressions here.

Reply:
https://github.com/snyk/snyk-ls/pull/234#discussion_r1131187779 
> moving config as a property of the scanner was less preferrable because the codeaction creation happens at the level of bundle, which has no reverse access to the scanner that created this bundle

maybe I'm wrong here, you mention that scanner is the singleton but I AFAIU accesing it goes via `snyk` package, then going to its product scanners, filtering them to find the code one etc -- just to arrive back to the code.

As opposed to that chain, keeping it just a config singleton at a package level allows to access it now (and in future) without overhead. Do I oversimplify or miss smth?

Reply:
I agree with the direction, although detaching the information stored in `Issue` from sounds like not a fully polished solution -- because it opens path to maintaining this map in sync with which issues are not dead (went out of scope) etc. 

You anyways already started with `any` there, so what I would do is (similar to scanners) to have an array of `[]any` in it, and each backend is then free to add its own struct:
```go
import "fmt"

type Issue struct {
	additionalData []any
}

type codeData struct {
	myData string
}

type ossData struct {
	myData int
}

func main() {
	issue := Issue{
		additionalData: []any{},
	}

	// code part creating issue
	issue.additionalData = append(issue.additionalData, codeData{myData: "code"})
	// oss part
	issue.additionalData = append(issue.additionalData, ossData{myData: 42})

	// part consuming the issue
	for _, data := range issue.additionalData {
		codeData, ok := data.(codeData)
		if !ok {
			continue
		}
		fmt.Print("Code data ", codeData.myData)
	}
}
```

All in all, this seems more conformant to the story of having product-specific things on top level (like scanners).